File: src/kiwimu/pipeline/graph.py
"""Build knowledge graph data for visualization."""

from __future__ import annotations

import json
from pathlib import Path

from ..store.store import Store


def build_graph_data(store: Store) -> dict:
    """Generate D3.js-compatible graph data from the links table."""
    pages = store.list_pages()
    links = store.get_all_links()

    # Calculate degree for node sizing
    degree: dict[int, int] = {}
    for page in pages:
        degree[page.id] = 0
    for link in links:
        degree[link.from_page_id] = degree.get(link.from_page_id, 0) + 1
        degree[link.to_page_id] = degree.get(link.to_page_id, 0) + 1

    nodes = [
        {
            "id": page.slug,
            "title": page.title,
            "degree": degree.get(page.id, 0),
        }
        for page in pages
    ]

    link_data = []
    slug_map = {p.id: p.slug for p in pages}
    for link in links:
        if link.from_page_id in slug_map and link.to_page_id in slug_map:
            link_data.append({
                "source": slug_map[link.from_page_id],
                "target": slug_map[link.to_page_id],
            })

    return {"nodes": nodes, "links": link_data}


def export_graph_json(store: Store, output_path: Path) -> None:
    """Write graph data to a JSON file."""
    data = build_graph_data(store)
    output_path.write_text(json.dumps(data, ensure_ascii=False, indent=2))

File: src/pipeline/graph.ts
import type { Store } from "../store";

export interface GraphData {
  nodes: Array<{ id: string; title: string; degree: number; type: string }>;
  links: Array<{ source: string; target: string }>;
}

export function buildGraphData(store: Store): GraphData {
  const pages = store.listPages();
  const links = store.getAllLinks();

  const degree = new Map<number, number>();
  for (const page of pages) degree.set(page.id, 0);
  for (const link of links) {
    degree.set(link.from_page_id, (degree.get(link.from_page_id) || 0) + 1);
    degree.set(link.to_page_id, (degree.get(link.to_page_id) || 0) + 1);
  }

  const slugMap = new Map(pages.map((p) => [p.id, p.slug]));

  return {
    nodes: pages.map((p) => ({
      id: p.slug,
      title: p.title,
      degree: degree.get(p.id) || 0,
      type: p.page_type,
    })),
    links: links
      .filter((l) => slugMap.has(l.from_page_id) && slugMap.has(l.to_page_id))
      .map((l) => ({
        source: slugMap.get(l.from_page_id)!,
        target: slugMap.get(l.to_page_id)!,
      })),
  };
}
